theano/gof/utils.py

```diff
@@ -210,6 +212,66 @@ def __str__(self):
 
         return type.__new__(cls, name, bases, dct)
 
+    _cache = {}
+
+    def __call__(cls, *args, **kwargs):
+        # Rules of caching:
+        #   - when the props is None, we never cache
+        #   - when the props is (), we always cache
+        #   - when the props is defined but not empty,
+        #     the args and kwargs should only contain the porps
+        # Note: when the props is empty inspect.getargspec(cls.__init__)
+        #       gives an error since cls.__init__ is not a python function.
+
+        props = cls.__dict__.get('__props__', None)
+
+        # The cases are being separated for code readability.
+        to_cache = False
+        if props is None:
+            pass
+        elif props == ():
+            to_cache = True
```

Comment:
I think this `elif props == ()` should be removed. The next one should cover that cases. If there is no props, there must be no parameter to the \__init__ to enable the cache. If there is parameter to the init and there is an empty props, then we should not cache.